dotnet/Utils/RichPath.cs

```diff
@@ -0,0 +1,439 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+using Newtonsoft.Json;
+using Microsoft.Azure.Storage.Auth;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+using System.IO.Compression;
+using System.Linq;
+
+namespace MSRC.DPU.Utils
+{
+    public abstract class RichPath
+    {
+        public const string AZURE_PATH_PREFIX = "azure://";
+
+        public static RichPath Create(string path, string azureInfoPath = null)
+        {
+            if (path.StartsWith(AZURE_PATH_PREFIX))
+            {
+                if (azureInfoPath == null)
+                {
+                    throw new ArgumentException($"Azure path {path} requires azure authentification info.");
+                }
+
+                // Strip off azure:// prefix:
+                var accountPath = path.Substring(AZURE_PATH_PREFIX.Length);
+                return new AzurePath(azureInfoPath, accountPath);
+            }
+            else
+            {
+                return new LocalPath(path);
+            }
+        }
+
+        public abstract string Path { get; }
+
+        public abstract bool IsDirectory();
+
+        public abstract bool IsFile();
+
+        public abstract void MakeAsDirectory();
+
+        public abstract Stream ReadAsBinaryPotentiallyUncompress();
+
+        public abstract void Store(object data);
+
+        public abstract IEnumerable<RichPath> FilesInDirectory(string searchPattern = "*");
+
+        public abstract RichPath Join(string newComponent);
+
+        protected static void StoreAsJSON(Stream outStream, object data, bool jsonl = false)
+        {
+            var serializer = new JsonSerializer { Formatting = Formatting.None }; // No whitespace, no linebreaks
+            using (var textWriter = new StreamWriter(outStream))
+            {
+                if (!jsonl)
+                {
+                    serializer.Serialize(textWriter, data);
+                }
+                else
+                {
+                    var collection = data as System.Collections.ICollection;
+                    foreach (var ele in collection)
```

Comment:
minor: maybe check if `collection == null` and throw an exception in that case?

Reply:
Done.